src/Tmds.Ssh/MultiSemaphore.cs

```diff
@@ -0,0 +1,88 @@
+// This file is part of Tmds.Ssh which is released under MIT.
+// See file LICENSE for full license details.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Tmds.Ssh
+{
+    sealed class MultiSemaphore : IDisposable
```

Comment:
Isn't there a Producer/Consumer collection implementation like https://docs.microsoft.com/cs-cz/dotnet/standard/collections/thread-safe/blockingcollection-overview that might do the same thing?

Reply:
Those need real elements. We track space available for send. We don't have something to store.